File: Problems/0500_0599/0538_Convert_BST_to_Greater_Tree/Project_CS/Convert_BST_to_Greater_Tree.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution
{
    public TreeNode ConvertBST(TreeNode root)
    {
        TreeNode node = root;
        Stack<TreeNode> stack = new Stack<TreeNode>();

        int sum = 0;
        while (node != null || stack.Count > 0)
        {
            if (node != null)
            {
                stack.Push(node);
                node = node.right;
            }
            else
            {
                // visit
                node = stack.Pop();
                sum += node.val;
                node.val = sum;
                
                node = node.left;
            }
        }
        
        return root;
    }
    
    List<int> nums = new List<int>();
    public TreeNode ConvertBST_work(TreeNode root)
    {
        getAllVal(root);

        nums.Sort();
        nums.Reverse();

        TreeNode resultNode = root;
        setGreater(resultNode);

        return resultNode;
    }

    private void getAllVal(TreeNode node)
    {
        if (node == null)
            return;
        nums.Add(node.val);
        if (node.left != null)
            getAllVal(node.left);
        if (node.right != null)
            getAllVal(node.right);
    }

    public void setGreater(TreeNode node)
    {
        if (node == null)
            return;
        node.val += calc_addVal(node.val);
        if (node.left != null)
            setGreater(node.left);
        if (node.right != null)
            setGreater(node.right);
    }

    private int calc_addVal(int currentVal)
    {
        int addVal = 0;
        for (int i = 0; i < nums.Count; ++i)
            if (nums[i] > currentVal)
                addVal += nums[i];
            else
                break;
        
        return addVal;
    }

    public void Main(string args)
    {
        Console.WriteLine("args = " + args );
        string flds = args.Replace("\"", "").Replace("[", "").Replace("]", "").Trim();

        OperateTreeNode ope_t = new OperateTreeNode();
        TreeNode root = ope_t.CreateTreeNode(flds);
        Console.Write("root = \n" + ope_t.TreeToStaircaseString(root));
        Console.WriteLine("root = \n" + ope_t.Tree2str(root));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        TreeNode result = ConvertBST(root);

        sw.Stop();
        Console.Write("result = \n" + ope_t.TreeToStaircaseString(result));
        Console.WriteLine("result = " + ope_t.Tree2str(result));
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0500_0599/0538_Convert_BST_to_Greater_Tree/Project_Python3/Convert_BST_to_Greater_Tree.py
# coding: utf-8

import os
import sys
import time

from TreeNode.Codec import Codec
from TreeNode.TreeNode import TreeNode
from TreeNode.OperateTreeNode import OperateTreeNode

class Solution:
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        inorder = []
        def traverse(root):
            if root is None:
                return
            else:
                traverse(root.left)
                inorder.append(root)
                traverse(root.right)
        traverse(root)
        n = len(inorder)
        for i in range(n-2, -1, -1):
            inorder[i].val += inorder[i+1].val
        return root
    
    def convertBST_work(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        self.nums = []
        self.getAllVal(root)
        self.nums.sort(reverse=True)
        self.setGreater(root)
        return root
  
    def getAllVal(self, node):
        if node is None:
            return
        self.nums.append(node.val)
        if node.left is not None:
            self.getAllVal(node.left)
        if node.right is not None:
            self.getAllVal(node.right)

    def setGreater(self, node):
        if node is None:
            return
        node.val += self.calc_addVal(node.val)
        if node.left is not None:
            self.setGreater(node.left)
        if node.right is not None:
            self.setGreater(node.right)
    
    def calc_addVal(self, currentVal):
        addVal = 0
        for i in range(len(self.nums)):
            if self.nums[i] > currentVal:
                addVal += self.nums[i]
            else:
                break
        return addVal

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python {0} <testdata.txt>".format(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("{0} not found...".format(argv[1]))
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = {0}".format(temp))
        loop_main(temp)
    #   print("Hit Return to continue...")
    #   input()

def loop_main(temp):
    flds = temp.replace("\"","").replace("[","").replace("]","").rstrip()

    ope_t = OperateTreeNode()
    root = ope_t.createTreeNode(flds)
    print("root = \n{0}".format(ope_t.treeToStaircaseString(root)))
    print("root = {0}".format(ope_t.tree2str(root)))

    sl = Solution()
    time0 = time.time()

    result = sl.convertBST(root)

    time1 = time.time()
    print("result = \n{0}".format(ope_t.treeToStaircaseString(result)))
    print("result = {0}".format(ope_t.tree2str(result)))
    print("Execute time ... : {0:f}[s]\n".format(time1 - time0))

if __name__ == "__main__":
    main()
